Introduce Promise wrapper for chrome API and setTimeout. #9

hidehikoabe opened the pull request:
More Promis-ify of gerrit.js the implementation as
reparation to introduce co.

hidehikoabe committed:
Introduce Promise wrapper for chrome API and setTimeout.

More Promis-ify of gerrit.js the implementation as
reparation to introduce co.

nya3jp commented on src/backends/gerrit.js:
```diff
@@ -37,54 +40,46 @@ export class GerritBackend {
       .then((text) => {
-        return JSON.parse(text.substring(text.indexOf('\n') + 1));
-      }).then((data) => {
+        let data = JSON.parse(text.substring(text.indexOf('\n') + 1))
```
nit: const

hidehikoabe replied:
Done.

nya3jp commented on src/chromeasync.js:
```diff
@@ -0,0 +1,59 @@
+
+/**
+ * Wraps a chrome API, which takes the callback with an argument at the last
+ * argument, * and returns a function returning the Promise instance.
```
nit: "\* and"

hidehikoabe replied:
Fixed.

nya3jp commented on src/chromeasync.js:
```diff
@@ -0,0 +1,59 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
```
Thanks for addition of this wrapper with clean JSDoc! This is definitely a good thing to have :)

nya3jp commented on src/chromeasync.js:
```diff
@@ -0,0 +1,59 @@
+ *     the given chrome API.
+ */
+function promisify(thisArg, name) {
+  let api = thisArg[name];
```
nit: const

hidehikoabe replied:
Done.

nya3jp commented on src/backends/gerrit.js:
```diff
@@ -12,17 +12,20 @@
 
   fetchAll_(selfAddress) {
-    var queries = [
+    let queries = [
```
Thanks for catching, I was not aware I still used var here. But could you use const if it's possible?

hidehikoabe replied:
Done.

nya3jp commented on src/chromeasync.js:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Wraps a chrome API, which takes the callback with an argument at the last
+ * argument, * and returns a function returning the Promise instance.
+ * Returned promise will be rejected if wrapped API succeeds (i.e.
```
Hmm, the explanation here sounds not true? I guess:

A returned promise will be resolved with the return value if the API call succeeds. Otherwise, it will be rejected with chrome.runtime.lastError value.

hidehikoabe replied:
Oops. That's my mistake. Fixed.

nya3jp commented on src/util.js:
```diff
@@ -35,3 +35,18 @@ export const formatPrettyTime = (timestamp) => {
+      reject(new Error('Nagative duration: ' + duration));
+      return;
+    }
+    setTimeout(duration, resolve);
```
I guess the argument order is swapped?

> var timeoutID = window.setTimeout(code, [delay]);

https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setTimeout

hidehikoabe replied:
Good catch! Fixed.

nya3jp commented on src/backends/gerrit.js:
```diff
@@ -103,14 +98,10 @@ export class GerritBackend {
-          }
-        });
-        if (reviewing) {
+        const owner_id = entry['owner']['_account_id'];
```
nit: could you use camelCase variable naming? i.e. ownerId.

hidehikoabe replied:
Done.

nya3jp commented on src/backends/gerrit.js:
```diff
@@ -37,54 +40,46 @@ export class GerritBackend {
-      const loginUrl = this.site_['url'] + '/login/';
-      chrome.tabs.create({url: loginUrl, active: true}, (tab) => {
+    const loginUrl = this.site_['url'] + '/login/';
+    return chromeAsync.tabs.create({url: loginUrl, active: true})
```
In the current implementation, the promise returned from login_() is resolved once login finishes. But IIUC in the new implementation it is resolved as soon as a new tab is created?

nya3jp commented:
Thanks for this change! My comments are all minor points so please feel free to commit once your address them (I think you have permission to do so, otherwise please let me know).

hidehikoabe committed:
Address review comments.

- Use const if possible.
- Update comments.
- Fix setTimeout arguments.
- Add import util.
- use camelCase for variable names.

hidehikoabe commented:
Thank you for review. Merging.
